bufferevent_ssl.c

```diff
@@ -268,20 +268,32 @@ do_read(struct bufferevent_ssl *bev_ssl, int n_to_read) {
 	if (n < 0)
 		return OP_ERR;
 
-	for (i=0; i<n; ++i) {
+	size_t len = 0;
+	for (i = 0; i < n;) {
 		if (bev_ssl->bev.read_suspended)
 			break;
 		bev_ssl->ssl_ops->clear_error();
-		r = bev_ssl->ssl_ops->read(bev_ssl->ssl, space[i].iov_base, space[i].iov_len);
-		if (r>0) {
+		r = bev_ssl->ssl_ops->read(
+			bev_ssl->ssl, space[i].iov_base + len, space[i].iov_len - len);
+		if (r > 0) {
 			result |= OP_MADE_PROGRESS;
 			if (bev_ssl->read_blocked_on_write)
 				if (clear_rbow(bev_ssl) < 0)
 					return OP_ERR | result;
-			++n_used;
-			space[i].iov_len = r;
 			bev_ssl->ssl_ops->decrement_buckets(bev_ssl);
+			len += r;
+			if (space[i].iov_len - len > 0) {
+				continue;
+			} else {
+				space[i].iov_len = len;
+				len = 0;
+				++i;
+			}
 		} else {
+			if (len > 0) {
+				space[i].iov_len = len;
+				++i;
+			}
 			int err = bev_ssl->ssl_ops->get_error(bev_ssl->ssl, r);
```

Comment:
/home/runner/work/libevent/libevent/dist/libevent-2.2.0-alpha-dev/bufferevent_ssl.c:297:4: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
  297 |    int err = bev_ssl->ssl_ops->get_error(bev_ssl->ssl, r);
      |    ^~~